### PR TITLE
#88 Cypress tests [WIP]

### PR DESCRIPTION
**Pull request by Vitaliy-1**

### CODE REVIEW

**Comment by Vitaliy-1** [comment]
Still a work in progress. Having some issues in selecting a color through a color picker with Cypress. @NateWr, does it also make sense to test rarely accessed pages, like subscription-related and authors search/index?

File: cypress/tests/functional/Classic.spec.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		'article/view/17/3',
+		'about',
+		'about/editorialTeam',
+		'about/submissions',
+		'about/contact',
+		'about/privacy',
+		'information/readers',
+		'information/authors',
+		'information/librarians'
+	];
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:28]: I recommend putting these directly into `cy.visit()` calls below. When the cypress tests error, you'll get the line of code in the error, but not necessarily the value of the variable. So the error message might not tell you which page failed.

File: cypress/tests/functional/Classic.spec.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		'familyName': 'Debreenik',
+		'username': 'jdebreenik',
+		'country': 'UA',
+		'affiliation': 'Lorem Ipsum University'
+	};
+
+	it('Enables and selects the theme', function() {
+		cy.login('admin', 'admin', 'publicknowledge');
+
+		cy.get('nav[class="app__nav"] ul a[class="app__navItem"]:contains("Website")').click();
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:47]: You'll save yourself some maintenance headaches by reducing the specificity of this selector:

```js
cy.get('.app__nav a:contains("Website")').click();
```

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  # Prepare OJS environment
+  - git clone -b ${BRANCH} https://github.com/pkp/${APPLICATION} ~/${APPLICATION}
+  - cd ~/${APPLICATION}
+  - git submodule update --init --recursive
+  - source lib/pkp/tools/travis/prepare-tests.sh
+  - lib/pkp/tools/travis/prepare-webserver.sh
+  # Build/install dependencies
+  - lib/pkp/tools/travis/install-composer-dependencies.sh
+  - npm i g -npm && npm install && npm run build
+  - ln -s ${TRAVIS_BUILD_DIR} ~/${APPLICATION}/plugins/themes/classic
```
Comment (NateWr) [inline .travis.yml:29]: Do you need to run any additional build tasks for the Classic theme?

File: cypress/tests/functional/Classic.spec.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		// Find and enable the plugin
+		cy.get('input[id^="select-cell-classicthemeplugin-enabled"]').click();
+		cy.get('div:contains(\'The plugin "Classic Theme" has been enabled.\')');
+		cy.reload();
+
+		// Select the Classic theme
+		cy.get('button[id="appearance-button"]').click();
+		cy.get('select[id="theme-themePluginPath-control"]').select('classic');
+		cy.get('div[id="theme"] button').contains('Save').click();
+		cy.get('div[id="theme"] [role="status"]').contains('Saved');
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:59]: Again, the less you specify the less likely you'll need to update the tests when there are syntax changes to the templates. Here I'd use:

```
cy.get('#theme [role="status"]').contains('Saved');
```

File: cypress/tests/functional/Classic.spec.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	});
+
+	it('Visits front-end theme pages', function() {
+		cy.visit(' ');
+		pages.forEach(page => {
+			cy.visit(path + '/' + page);
+		});
+	});
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:67]: If you have the time and energy, I'd spot check some of the data on key pages. Some ideas:

- Are all article titles in the issue there? 
- Any article titles showing that shouldn't?
- Is the correct version of a multi-version article shown on issue toc, article landing page, etc?

File: cypress/tests/functional/Classic.spec.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		cy.get('select[name="dateFromDay"]').select('1');
+		cy.get('select[name="dateToYear"]').select(year);
+		cy.get('select[name="dateToMonth"]').select(month);
+		cy.get('select[name="dateToDay"]').select(day);
+
+		cy.get('input[name="authors"]').type('Vajiheh Karbasizaed', {delay: 0});
+		cy.get('button[type="submit"]').click();
+		cy.get('.search_results').children().should('have.length', 1);
+		cy.get('.article_summary').first().click();
+		cy.url().should('match', /article\/view/);
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:85]: Would this catch a case where the URL is 404? Maybe check the `<h1>` for the correct title?

File: cypress/tests/functional/Classic.spec.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	});
+
+	it('Register a user', function() {
+		// Sign out
+		cy.visit(path + '/' + 'login/signOut');
+		cy.url().should('match', /login/);
+
+		// Register; 'register' command won't work for this theme because privacyConsent label overlays input checkbox
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:93]: Is this comment still relevant?

File: cypress/tests/functional/Classic.spec.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		cy.get('#affiliation').type(user.affiliation, {delay: 0});
+		cy.get('#country').select(user.country);
+		cy.get('#username').type(user.username, {delay: 0});
+		cy.get('#email').type(user.username + '@mailinator.com', {delay: 0});
+		cy.get('#password').type(user.username + user.username, {delay: 0});
+		cy.get('#password2').type(user.username + user.username, {delay: 0});
+		cy.get('label[for="privacyConsent"]').click();
+		cy.get('label[for="checkbox-reviewer-interests"]').click();
+		cy.get('#tagitInput input').type('psychotherapy,neuroscience,neurobiology', {delay: 0});
+		cy.get('button[type="submit"]').contains('Register').click().click(); // Cypress expects 2 clicks to submit the form
```
Comment (NateWr) [inline cypress/tests/functional/Classic.spec.js:107]: Probably the first click is bluring the `tagit` or something...

**Review by NateWr — COMMENTED** [review]
Looks great! This will be really useful to have on our themes. 

> Having some issues in selecting a color through a color picker with Cypress.

```js
cy.get('input.vc-input__input').clear().type('#AAAAAA');
```

If that's not working, you could always send a request directly to the REST API.

> does it also make sense to test rarely accessed pages, like subscription-related and authors search/index?

Yes, if you have the time and energy and the theme supports them. Rarely accessed pages are even more likely to have undiscovered bugs.

Other config variables to test, _if you have the time/energy_:

- With/without journal logo
- Homepage display options
- Category pages
- Article cover images
- Issue descriptions/covers/galleys
- Custom section/article orders in issues
- Mulutilingual support
- Galleys that require login/payment

(A bunch of these aren't even tested in the default theme!)

File: cypress/tests/functional/Classic.spec.js (hunk 7 above)
Comment (Vitaliy-1) [reply cypress/tests/functional/Classic.spec.js:93]: Yes. I meant here `cy.register()`, one of the custom commands that I found is added.

File: .travis.yml (hunk 3 above)
Comment (Vitaliy-1) [reply .travis.yml:29]: The theme has only dev dependencies. If anyone wants to update,e.g., Bootstrap, I'm assuming it's done locally with Gulp tasks before making a PR.

File: cypress/tests/functional/Classic.spec.js (hunk 4 above)
Comment (Vitaliy-1) [reply cypress/tests/functional/Classic.spec.js:44]: Interesting that this line sometimes fails locally. This happens the first time this script is run after OJS installation, I've restarted the test to see if it's the case for Travis integration too.

Comment (Vitaliy-1) [reply cypress/tests/functional/Classic.spec.js:44]: My mistake, this one: `cy.get('div:contains(\'The plugin "Classic Theme" has been enabled.\')')` after theme plugin is enabled

**Pull request merged**